File: EasyDarkMode+Remote/UIButton+DMRemote.m
//
//  UIButton+DMRemote.m
//  EasyDarkMode
//
//  Created by Elenion on 2019/8/21.
//  Copyright © 2019 Elenion. All rights reserved.
//

#import "UIButton+DMRemote.h"
#import "UIImage+DMRemote.h"
#import "UIImage+DM.h"

@implementation UIButton (DMRemote)

- (void)dm_setImageUrlLight:(NSURL *)light dark:(NSURL *)dark forState:(UIControlState)state {
    __weak __auto_type selfWeak = self;
    [UIImage dm_imageWithUrlLight:light dark:dark completion:^(UIImage * _Nullable image, NSError * _Nullable error) {
        [selfWeak setImage:image forState:state];
    }];
}

- (void)dm_setImagePathStringLight:(NSString *)light dark:(NSString *)dark forState:(UIControlState)state {
    NSURL *lightUrl = light ? [NSURL URLWithString:light] : nil;
    NSURL *darkUrl = dark ? [NSURL URLWithString:dark] : nil;
    [self dm_setImageUrlLight:lightUrl dark:darkUrl forState:state];
}

- (void)dm_setBackgroundUrlLight:(NSURL *)light dark:(NSURL *)dark forState:(UIControlState)state {
    __weak __auto_type selfWeak = self;
    [UIImage dm_imageWithUrlLight:light dark:dark completion:^(UIImage * _Nullable image, NSError * _Nullable error) {
        [selfWeak setBackgroundImage:image forState:state];
    }];
}

- (void)dm_setBackgroundPathStringLight:(NSString *)light dark:(NSString *)dark forState:(UIControlState)state {
    NSURL *lightUrl = light ? [NSURL URLWithString:light] : nil;
    NSURL *darkUrl = dark ? [NSURL URLWithString:dark] : nil;
    [self dm_setBackgroundUrlLight:lightUrl dark:darkUrl forState:state];
}

@end

File: EasyDarkMode+Remote+Swift/UIButton+DMRemote.swift
//
//  UIButton+DMRemote.swift
//  EasyDarkMode
//
//  Created by Elenion on 2019/8/22.
//  Copyright © 2019 Elenion. All rights reserved.
//

import UIKit

extension UIButton: InterfaceCompatable {
    
}

public extension InterfaceAdaptor where Base: UIButton {
    func setImagePathUrl(light: URL, dark: URL, for state: UIControl.State) {
        base.dm_setImageUrlLight(light, dark: dark, for: state)
    }
    
    func setImagePathString(light: String, dark: String, for state: UIControl.State) {
        base.dm_setImagePathStringLight(light, dark: dark, for: state)
    }
    
    func setBackgroundImagePathUrl(light: URL, dark: URL, for state: UIControl.State) {
        base.dm_setBackgroundUrlLight(light, dark: dark, for: state)
    }
    
    func setBackgroundImagePathString(light: String, dark: String, for state: UIControl.State) {
        base.dm_setBackgroundPathStringLight(light, dark: dark, for: state)
    }
}

public extension InterfaceAdaptor where Base: UIButton {
    @available(*, deprecated, renamed: "setImagePathUrl(light:dark:for:)")
    func setImage(url light: URL?, dark: URL?, for state: UIControl.State) {
        guard let light = light, let dark = dark else {
            return
        }
        setImagePathUrl(light: light, dark: dark, for: state)
    }
    
    @available(*, deprecated, renamed: "setImagePathString(light:dark:for:)")
    func setImage(pathString light: String?, dark: String?, for state: UIControl.State) {
        guard let light = light, let dark = dark else {
            return
        }
        setImagePathString(light: light, dark: dark, for: state)
    }
    
    @available(*, deprecated, renamed: "setBackgroundImagePathUrl(light:dark:for:)")
    func setBackgroundImage(url light: URL?, dark: URL?, for state: UIControl.State) {
        guard let light = light, let dark = dark else {
            return
        }
        setBackgroundImagePathUrl(light: light, dark: dark, for: state)
    }
    
    @available(*, deprecated, renamed: "setBackgroundImagePathString(light:dark:for:)")
    func setBackgroundImage(pathString light: String?, dark: String?, for state: UIControl.State) {
        guard let light = light, let dark = dark else {
            return
        }
        setBackgroundImagePathString(light: light, dark: dark, for: state)
    }
}
